File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/flutter/VideoTrackBridge.kt
package com.theoplayer.flutter

import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.track.mediatrack.video.ActiveQualityChangedEvent
import com.theoplayer.android.api.event.track.mediatrack.video.VideoTrackEventTypes
import com.theoplayer.android.api.event.track.mediatrack.video.TargetQualityChangedEvent
import com.theoplayer.android.api.event.track.mediatrack.video.list.AddTrackEvent
import com.theoplayer.android.api.event.track.mediatrack.video.list.VideoTrackListEventTypes
import com.theoplayer.android.api.event.track.mediatrack.video.list.RemoveTrackEvent
import com.theoplayer.android.api.event.track.mediatrack.video.list.TrackListChangeEvent
import com.theoplayer.android.api.event.track.mediatrack.quality.QualityEventTypes
import com.theoplayer.android.api.event.track.mediatrack.quality.QualityUpdateEvent
import com.theoplayer.android.api.player.Player
import com.theoplayer.android.api.player.track.mediatrack.MediaTrack
import com.theoplayer.android.api.player.track.mediatrack.quality.VideoQuality
import com.theoplayer.flutter.pigeon.THEOplayerFlutterVideoTracksAPI
import com.theoplayer.flutter.pigeon.THEOplayerNativeVideoTracksAPI
import com.theoplayer.flutter.pigeon.THEOplayerNativeVideoTracksAPI.Companion.setUp

class VideoTrackBridge(
    private val player: Player,
    pigeonMessenger: PigeonBinaryMessengerWrapper,
) : THEOplayerNativeVideoTracksAPI {

    private val flutterVideoTracksAPI = THEOplayerFlutterVideoTracksAPI(pigeonMessenger)

    init {
        setUp(pigeonMessenger, this)
    }

    private val emptyCallback: (Result<Unit>) -> Unit = {}

    private val addVideoTrackListener: EventListener<AddTrackEvent> = EventListener {
        flutterVideoTracksAPI.onAddVideoTrack(
            it.track.id,
            it.track.uid.toLong(),
            it.track.label,
            it.track.language,
            it.track.kind,
            it.track.isEnabled,
            emptyCallback
        )

        it.track.qualities?.forEach { quality ->
            flutterVideoTracksAPI.onVideoTrackAddQuality(
                it.track.uid.toLong(),
                quality.id,
                quality.uid.toLong(),
                quality.name,
                quality.bandwidth,
                quality.codecs,
                quality.width.toLong(),
                quality.height.toLong(),
                quality.frameRate,
                quality.firstFrame,
                emptyCallback
            )
        }

        attachTrackListeners(it.track)
    }

    private val removeVideoTrackListener = EventListener<RemoveTrackEvent> {
        removeTrackListeners(it.track)
        flutterVideoTracksAPI.onRemoveVideoTrack(it.track.uid.toLong(), emptyCallback)
    }

    private val videoTrackListChangeListener = EventListener<TrackListChangeEvent> {
        flutterVideoTracksAPI.onVideoTrackListChange(it.track.uid.toLong(), emptyCallback)
    }

    private fun videoTrackTargetQualityChangedListener(track: MediaTrack<VideoQuality>) = EventListener<TargetQualityChangedEvent> {
        flutterVideoTracksAPI.onTargetQualityChange(
            track.uid.toLong(),
            it.qualities.map { videoQuality -> videoQuality.uid.toLong() },
            it.quality?.uid?.toLong(),
            emptyCallback
        )
    }

    private fun videoTrackActiveQualityChangedListener(track: MediaTrack<VideoQuality>) = EventListener<ActiveQualityChangedEvent> {
        flutterVideoTracksAPI.onActiveQualityChange(track.uid.toLong(), it.quality.uid.toLong(), emptyCallback)
    }

    private fun videoQualityChangeListener(track: MediaTrack<VideoQuality>, quality: VideoQuality) = EventListener<QualityUpdateEvent> {
        flutterVideoTracksAPI.onQualityUpdate(
            track.uid.toLong(),
            quality.uid.toLong(),
            quality.name,
            quality.bandwidth,
            quality.codecs,
            quality.width.toLong(),
            quality.height.toLong(),
            quality.frameRate,
            quality.firstFrame,
            emptyCallback
        )
    }

    fun attachListeners() {
        player.videoTracks.addEventListener(VideoTrackListEventTypes.ADDTRACK, addVideoTrackListener)
        player.videoTracks.addEventListener(VideoTrackListEventTypes.REMOVETRACK, removeVideoTrackListener)
        player.videoTracks.addEventListener(VideoTrackListEventTypes.TRACKLISTCHANGE, videoTrackListChangeListener)
    }

    fun removeListeners() {
        player.videoTracks.removeEventListener(VideoTrackListEventTypes.ADDTRACK, addVideoTrackListener)
        player.videoTracks.removeEventListener(VideoTrackListEventTypes.REMOVETRACK, removeVideoTrackListener)
        player.videoTracks.removeEventListener(VideoTrackListEventTypes.TRACKLISTCHANGE, videoTrackListChangeListener)
        player.videoTracks.forEach { track ->
            removeTrackListeners(track)
        }
    }

    private fun attachTrackListeners(track: MediaTrack<VideoQuality>) {
        track.addEventListener(VideoTrackEventTypes.TARGETQUALITYCHANGEDEVENT, videoTrackTargetQualityChangedListener(track))
        track.addEventListener(VideoTrackEventTypes.ACTIVEQUALITYCHANGEDEVENT, videoTrackActiveQualityChangedListener(track))
        track.qualities?.forEach {
            attachQualityListeners(track, it)
        }
    }

    private fun removeTrackListeners(track: MediaTrack<VideoQuality>) {
        track.removeEventListener(VideoTrackEventTypes.TARGETQUALITYCHANGEDEVENT, videoTrackTargetQualityChangedListener(track))
        track.removeEventListener(VideoTrackEventTypes.ACTIVEQUALITYCHANGEDEVENT, videoTrackActiveQualityChangedListener(track))
        track.qualities?.forEach {
            removeQualityListeners(track, it)
        }
    }

    private fun attachQualityListeners(track: MediaTrack<VideoQuality>, quality: VideoQuality) {
        quality.addEventListener(QualityEventTypes.UPDATE, videoQualityChangeListener(track, quality))
    }

    private fun removeQualityListeners(track: MediaTrack<VideoQuality>, quality: VideoQuality) {
        quality.removeEventListener(QualityEventTypes.UPDATE, videoQualityChangeListener(track, quality))
    }

    override fun setTargetQuality(videoTrackUid: Long, qualityUid: Long?) {
        val videoTrack = player.videoTracks.find { it.uid.toLong() == videoTrackUid }
        val videoQuality = videoTrack?.qualities?.find { it.uid.toLong() == qualityUid }
        videoTrack?.targetQuality = videoQuality
    }

    override fun setTargetQualities(videoTrackUid: Long, qualitiesUid: List<Long>?) {
        val videoTrack = player.videoTracks.find { it.uid.toLong() == videoTrackUid }
        val videoQualities = videoTrack?.qualities?.filter { qualitiesUid?.contains(it.uid.toLong()) == true }
        videoTrack?.setTargetQualities(videoQualities)
    }

    override fun setEnabled(videoTrackUid: Long, enabled: Boolean) {
        val videoTrack = player.videoTracks.find { it.uid.toLong() == videoTrackUid }
        videoTrack?.isEnabled = enabled
    }

}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/VideoTrackBridge.swift
//
//  VideoTrackBridge.swift
//  flutter_theoplayer_sdk_ios
//
//  Created by Hovig on 27/11/2023.
//

import Foundation
import THEOplayerSDK
import Flutter

class VideoTrackBridge: THEOplayerNativeVideoTracksAPI {

    private let theoplayer: THEOplayer
    private let pigeonMessenger: PigeonBinaryMessengerWrapper
    private let flutterVideoTracksAPI: THEOplayerFlutterVideoTracksAPI

    private var addVideoTrackListener: EventListener?
    private var removeVideoTrackListener: EventListener?
    private var videoTrackListChangeListener: EventListener?
    private var trackEventListeners: [Int: [EventListener]] = [:]

    private let emptyCompletion: (Result<Void, PigeonError>) -> Void = {result in }

    init(theoplayer: THEOplayer, pigeonMessenger: PigeonBinaryMessengerWrapper) {
        self.theoplayer = theoplayer
        self.pigeonMessenger = pigeonMessenger
        self.flutterVideoTracksAPI = THEOplayerFlutterVideoTracksAPI(binaryMessenger: pigeonMessenger)
        THEOplayerNativeVideoTracksAPISetup.setUp(binaryMessenger: pigeonMessenger, api: self)
    }

    func attachListeners() {
        addVideoTrackListener = theoplayer.videoTracks.addEventListener(type: VideoTrackListEventTypes.ADD_TRACK, listener: { [weak self] event in
            guard let self else { return }
            if let videoTrack = event.track as? VideoTrack {
                self.flutterVideoTracksAPI.onAddVideoTrack(
                    id: videoTrack.id,
                    uid: Int64(videoTrack.uid),
                    label: videoTrack.label,
                    language: videoTrack.language,
                    kind: videoTrack.kind,
                    isEnabled: videoTrack.enabled,
                    unlocalizedLabel: videoTrack.unlocalizedLabel,
                    completion: self.emptyCompletion)

                // Emit quality info for each quality in the track
                let qualities = videoTrack.qualities
                for i in 0..<qualities.count {
                    let quality = qualities.get(i)
                    let identifier = String(quality.bandwidth)
                    var width: Int64 = 0
                    var height: Int64 = 0
                    if let videoQuality = quality as? VideoQuality {
                        width = Int64(videoQuality.width)
                        height = Int64(videoQuality.height)
                    }
                    self.flutterVideoTracksAPI.onVideoTrackAddQuality(
                        videoTrackUid: Int64(videoTrack.uid),
                        qualityId: identifier,
                        qualityUid: Int64(quality.bandwidth),
                        name: self.labelFromBandwidth(quality.bandwidth),
                        bandwidth: Int64(quality.bandwidth),
                        codecs: "",
                        width: width,
                        height: height,
                        frameRate: 0,
                        firstFrame: 0,
                        averageBandwidth: quality.averageBandwidth.map { Int64($0) },
                        available: true,
                        completion: self.emptyCompletion)
                }

                self.attachTrackListeners(videoTrack: videoTrack)
            }
        })

        removeVideoTrackListener = theoplayer.videoTracks.addEventListener(type: VideoTrackListEventTypes.REMOVE_TRACK, listener: { [weak self] event in
            guard let self else { return }
            self.removeTrackListeners(trackUid: event.track.uid)
            self.flutterVideoTracksAPI.onRemoveVideoTrack(uid: Int64(event.track.uid), completion: self.emptyCompletion)
        })

        videoTrackListChangeListener = theoplayer.videoTracks.addEventListener(type: VideoTrackListEventTypes.CHANGE, listener: { [weak self] event in
            guard let self else { return }
            self.flutterVideoTracksAPI.onVideoTrackListChange(uid: Int64(event.track.uid), completion: self.emptyCompletion)
        })
    }

    private func attachTrackListeners(videoTrack: VideoTrack) {
        var listeners: [EventListener] = []

        let activeQualityChangedListener = videoTrack.addEventListener(type: MediaTrackEventTypes.ACTIVE_QUALITY_CHANGED, listener: { [weak self] (event: ActiveQualityChangedEvent) in
            guard let self else { return }
            self.flutterVideoTracksAPI.onActiveQualityChange(
                videoTrackUid: Int64(videoTrack.uid),
                qualityUid: Int64(event.quality.bandwidth),
                completion: self.emptyCompletion)
        })
        listeners.append(activeQualityChangedListener)

        trackEventListeners[videoTrack.uid] = listeners
    }

    private func removeTrackListeners(trackUid: Int) {
        trackEventListeners.removeValue(forKey: trackUid)
    }

    private func removeListeners() {
        if let listener = addVideoTrackListener {
            theoplayer.videoTracks.removeEventListener(type: VideoTrackListEventTypes.ADD_TRACK, listener: listener)
        }
        if let listener = removeVideoTrackListener {
            theoplayer.videoTracks.removeEventListener(type: VideoTrackListEventTypes.REMOVE_TRACK, listener: listener)
        }
        if let listener = videoTrackListChangeListener {
            theoplayer.videoTracks.removeEventListener(type: VideoTrackListEventTypes.CHANGE, listener: listener)
        }
        trackEventListeners.removeAll()
    }

    func dispose() {
        removeListeners()
        THEOplayerNativeVideoTracksAPISetup.setUp(binaryMessenger: pigeonMessenger, api: nil)
    }

    func setTargetQuality(videoTrackUid: Int64, qualityUid: Int64?) throws {
        for i in 0..<theoplayer.videoTracks.count {
            let track = theoplayer.videoTracks.get(i)
            if track.uid == Int(videoTrackUid) {
                if let qualityUid = qualityUid {
                    let qualities = track.qualities
                    for j in 0..<qualities.count {
                        let quality = qualities.get(j)
                        if quality.bandwidth == Int(qualityUid) {
                            track.targetQualities = [quality]
                            return
                        }
                    }
                } else {
                    track.targetQualities = nil
                }
                return
            }
        }
    }

    func setTargetQualities(videoTrackUid: Int64, qualitiesUid: [Int64]?) throws {
        for i in 0..<theoplayer.videoTracks.count {
            let track = theoplayer.videoTracks.get(i)
            if track.uid == Int(videoTrackUid) {
                if let qualitiesUid = qualitiesUid {
                    var targetQualities: [any Quality] = []
                    let qualities = track.qualities
                    for j in 0..<qualities.count {
                        let quality = qualities.get(j)
                        if qualitiesUid.contains(Int64(quality.bandwidth)) {
                            targetQualities.append(quality)
                        }
                    }
                    track.targetQualities = targetQualities
                } else {
                    track.targetQualities = nil
                }
                return
            }
        }
    }

    func setEnabled(videoTrackUid: Int64, enabled: Bool) throws {
        for i in 0..<theoplayer.videoTracks.count {
            var track = theoplayer.videoTracks.get(i)
            if (track.uid == videoTrackUid) {
                track.enabled = enabled
            }
        }
    }

    private func labelFromBandwidth(_ bandwidth: Int) -> String {
        if bandwidth >= 1_000_000 {
            return String(format: "%.1f Mbps", Double(bandwidth) / 1_000_000.0)
        } else {
            return String(format: "%d kbps", bandwidth / 1000)
        }
    }

}
